File: 04_Unsupervised_ML_Examples/Scripts_R/SCRIPT - Países.R

# Análise Fatorial PCA + Análise de Clusters Hierárquico

# Curso: MBA DSA (USP ESALQ)

# Prof. Wilson Tarantin Jr.

# Fonte dos dados: https://www.kaggle.com/datasets/vipulgohel/clustering-pca-assignment?resource=download&select=Country-data.csv

# Instalação e carregamento dos pacotes utilizados

pacotes <- c("plotly",
             "tidyverse",
             "ggrepel",
             "knitr", "kableExtra",
             "reshape2",
             "PerformanceAnalytics",
             "psych",
             "Hmisc",
             "readxl",
             "cluster",
             "factoextra") 

if(sum(as.numeric(!pacotes %in% installed.packages())) != 0){
  instalador <- pacotes[!pacotes %in% installed.packages()]
  for(i in 1:length(instalador)) {
    install.packages(instalador, dependencies = T)
    break()}
  sapply(pacotes, require, character = T) 
} else {
  sapply(pacotes, require, character = T) 
}

# Carregamento da base de dados
paises <- read.csv("Países PCA Cluster.csv", sep = ",", dec = ".")

# Estatísticas descritivas
summary(paises)

# Scatter e ajuste entre as variáveis 'renda' e 'expectativa de vida'
paises %>%
  ggplot() +
  geom_point(aes(x = income, y = life_expec),
             color = "darkorchid",
             size = 3) +
  geom_smooth(aes(x = income, y = life_expec),
              color = "orange", 
              method = "loess", 
              formula = y ~ x, 
              se = FALSE,
              size = 1.3) +
  labs(x = "income",
       y = "life_expec") +
  theme_bw()

# Scatter e ajuste entre as variáveis 'exports' e 'gdpp'
paises %>%
  ggplot() +
  geom_point(aes(x = exports, y = gdpp),
             color = "darkorchid",
             size = 3) +
  geom_smooth(aes(x = exports, y = gdpp),
              color = "orange", 
              method = "loess", 
              formula = y ~ x, 
              se = FALSE,
              size = 1.3) +
  labs(x = "exports",
       y = "gdpp") +
  theme_bw()

# Coeficientes de correlação de Pearson para cada par de variáveis
rho <- rcorr(as.matrix(paises[,2:10]), type="pearson")

correl <- rho$r # Matriz de correlações
sig_correl <- round(rho$P, 4) # Matriz com p-valor dos coeficientes

# Elaboração de um mapa de calor das correlações de Pearson entre as variáveis
ggplotly(
  paises[,2:10] %>%
    cor() %>%
    melt() %>%
    rename(Correlação = value) %>%
    ggplot() +
    geom_tile(aes(x = Var1, y = Var2, fill = Correlação)) +
    geom_text(aes(x = Var1, y = Var2, label = format(Correlação, digits = 1)),
              size = 5) +
    scale_fill_viridis_b() +
    labs(x = NULL, y = NULL) +
    theme_bw())

# Visualização das distribuições das variáveis, scatters, valores das correlações
chart.Correlation(paises[,2:10], histogram = TRUE, pch = "+")

### Elaboração a Análise Fatorial Por Componentes Principais ###

# Teste de esfericidade de Bartlett
cortest.bartlett(paises[,2:10])

# Elaboração da análise fatorial por componentes principais
fatorial <- principal(paises[,2:10],
                      nfactors = length(paises[,2:10]),
                      rotate = "none",
                      scores = TRUE)
fatorial

# Eigenvalues (autovalores)
eigenvalues <- round(fatorial$values, 5)
eigenvalues

# Soma dos eigenvalues = 9 (quantidade de variáveis na análise)
# Também representa a quantidade máxima de possíveis fatores na análise
round(sum(eigenvalues), 2)

# Identificação da variância compartilhada em cada fator
variancia_compartilhada <- as.data.frame(fatorial$Vaccounted) %>% 
  slice(1:3)

rownames(variancia_compartilhada) <- c("Autovalores",
                                       "Prop. da Variância",
                                       "Prop. da Variância Acumulada")

# Variância compartilhada pelas variáveis originais para a formação de cada fator
round(variancia_compartilhada, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

# Cálculo dos scores fatoriais
scores_fatoriais <- as.data.frame(fatorial$weights)

# Visualização dos scores fatoriais
round(scores_fatoriais, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

# Cálculo dos fatores propriamente ditos
fatores <- as.data.frame(fatorial$scores)

View(fatores)

# Coeficientes de correlação de Pearson para cada par de fatores (ortogonais)
rho <- rcorr(as.matrix(fatores), type="pearson")
round(rho$r, 4)

# Cálculo das cargas fatoriais
cargas_fatoriais <- as.data.frame(unclass(fatorial$loadings))

# Visualização das cargas fatoriaisX
round(cargas_fatoriais, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

# Cálculo das comunalidades
comunalidades <- as.data.frame(unclass(fatorial$communality)) %>%
  rename(comunalidades = 1)

# Visualização das comunalidades (aqui são iguais a 1 para todas as variáveis)
# Foram extraídos 9 fatores neste primeiro momento
round(comunalidades, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped",
                full_width = FALSE,
                font_size = 20)

### Elaboração da Análise Fatorial por Componentes Principais ###
### Fatores extraídos a partir de autovalores maiores que 1 ###

# Definição da quantidade de fatores com eigenvalues maiores que 1
k <- sum(eigenvalues > 1)
print(k)

# Elaboração da análise fatorial por componentes principais sem rotação
# Com quantidade 'k' de fatores com eigenvalues maiores que 1
fatorial2 <- principal(paises[,2:10],
                       nfactors = k,
                       rotate = "none",
                       scores = TRUE)
fatorial2

# Cálculo das comunalidades com apenas os 'k' ('k' = 3) primeiros fatores
comunalidades2 <- as.data.frame(unclass(fatorial2$communality)) %>%
  rename(comunalidades = 1)

# Visualização das comunalidades com apenas os 'k' ('k' = 3) primeiros fatores
round(comunalidades2, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped",
                full_width = FALSE,
                font_size = 20)

# Loading plot com as cargas dos dois primeiros fatores
cargas_fatoriais[, 1:2] %>% 
  data.frame() %>%
  rownames_to_column("variáveis") %>%
  ggplot(aes(x = PC1, y = PC2, label = variáveis)) +
  geom_point(color = "darkorchid",
             size = 3) +
  geom_text_repel() +
  geom_vline(aes(xintercept = 0), linetype = "dashed", color = "orange") +
  geom_hline(aes(yintercept = 0), linetype = "dashed", color = "orange") +
  expand_limits(x= c(-1.25, 0.25), y=c(-0.25, 1)) +
  theme_bw()

# Adicionando os fatores extraídos no banco de dados original
paises <- bind_cols(paises,
                    "fator_1" = fatores$PC1, 
                    "fator_2" = fatores$PC2,
                    "fator_3" = fatores$PC3)

# Análise de Cluster Utilizando os 3 Fatores

# Análise dos fatores (média e desvio padrão)
summary(paises[,11:13])
sd(paises[,11])
sd(paises[,12])
sd(paises[,13])

## ATENÇÃO: os clusters serão formados a partir dos 3 fatores
## Não aplicaremos o Z-Score, pois os fatores já são padronizados

# Matriz de dissimilaridades
matriz_D <- paises[,11:13] %>% 
  dist(method = "euclidean")

# Elaboração da clusterização hierárquica
cluster_hier <- agnes(x = matriz_D, method = "complete")

# Definição do esquema hierárquico de aglomeração

# As distâncias para as combinações em cada estágio
coeficientes <- sort(cluster_hier$height, decreasing = FALSE) 
coeficientes

# Tabela com o esquema de aglomeração. Interpretação do output:

## As linhas são os estágios de aglomeração
## Nas colunas Cluster1 e Cluster2, observa-se como ocorreu a junção
## Quando for número negativo, indica observação isolada
## Quando for número positivo, indica cluster formado anteriormente (estágio)
## Coeficientes: as distâncias para as combinações em cada estágio

esquema <- as.data.frame(cbind(cluster_hier$merge, coeficientes))
names(esquema) <- c("Cluster1", "Cluster2", "Coeficientes")
esquema

# Construção do dendrograma
dev.off()
fviz_dend(x = cluster_hier, show_labels = FALSE)

# Dendrograma com visualização dos clusters (definição de 10 clusters)
fviz_dend(x = cluster_hier,
          h = 3.0,
          show_labels = FALSE,
          color_labels_by_k = F,
          rect = F,
          rect_fill = F,
          ggtheme = theme_bw())

# Criando variável categórica para indicação do cluster no banco de dados
## O argumento 'k' indica a quantidade de clusters
paises$cluster_H <- factor(cutree(tree = cluster_hier, k = 10))

# Análise de variância de um fator (ANOVA). Interpretação do output:

## Mean Sq do cluster_H: indica a variabilidade entre grupos
## Mean Sq dos Residuals: indica a variabilidade dentro dos grupos
## F value: estatística de teste (Sum Sq do cluster_H / Sum Sq dos Residuals)
## Pr(>F): p-valor da estatística 
## p-valor < 0.05: pelo menos um cluster apresenta média estatisticamente diferente dos demais

## A variável mais discriminante dos grupos contém maior estatística F (e significativa)

# ANOVA da variável 'fator 1'
summary(anova_fator_1 <- aov(formula = fator_1 ~ cluster_H,
                             data = paises))

# ANOVA da variável 'fator 2'
summary(anova_fator_2 <- aov(formula = fator_2 ~ cluster_H,
                             data = paises))

# ANOVA da variável 'fator 3'
summary(anova_fator_3 <- aov(formula = fator_3 ~ cluster_H,
                             data = paises))

# Algumas estatísticas descritivas entre clusters

# PIB
group_by(paises, cluster_H) %>%
  summarise(
    mean = mean(gdpp, na.rm = TRUE),
    sd = sd(gdpp, na.rm = TRUE),
    min = min(gdpp, na.rm = TRUE),
    max = max(gdpp, na.rm = TRUE),
    obs = n())

# Indicador de saúde
group_by(paises, cluster_H) %>%
  summarise(
    mean = mean(health, na.rm = TRUE),
    sd = sd(health, na.rm = TRUE),
    min = min(health, na.rm = TRUE),
    max = max(health, na.rm = TRUE),
    obs = n())

# Expectativa de vida
group_by(paises, cluster_H) %>%
  summarise(
    mean = mean(life_expec, na.rm = TRUE),
    sd = sd(life_expec, na.rm = TRUE),
    min = min(life_expec, na.rm = TRUE),
    max = max(life_expec, na.rm = TRUE),
    obs = n())

# FIM!

File: 04_Unsupervised_ML_Examples/Scripts_Python/SCRIPT - Países.py
# Geral
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# PCA
from scipy.stats import pearsonr
import scipy.linalg as sla
from scipy.stats import zscore
import prince


# Clustering
from scipy.spatial.distance import pdist, squareform
import scipy.cluster.hierarchy as sch
import statsmodels.api as sm
from statsmodels.formula.api import ols



# PCA
# 01º Passo: Matriz de correlações de Pearson
# 02° Passo: Teste de esferecidade de Barllet
# 03° Passo: Autovalores e autovetores da matriz de correlação de Pearson. Da pra tirar a variância compartilhada, que é a porcentagem de cada autovalor da soma deles
# 04° Passo: Scores fatoriais (autovetores dividido pela raiz do seu autovalor)
# 05° Passo: Zscores (são as observações padronizadas)
# 06° Passo: Fatores (multiplicação matricial dos Zscores com os scores fatoriais)
# 07° Passo: Cargas fatoriais (correlaçaõ de Pearson entre os fatores e as variáveis originais)
# 08° Passo: Critério de Kaiser (autovalores maiores que 1), para definir o número de fatores
# 09° Passo: Comunalidades dos autovalores de Kaiser
# 10° Passo: Plotar as cargas fatoriais

# Clustering Hierarquico
# 01° Passo: Padronização das variáveis com o ZScore
# 02° Passo: Matriz de dissimilaridades    
# 03° Passo: Clusterização
# 04° Passo: Dendograma
# 05° Passo: Definir número de clusters e adicioná-los no dataframe
# 06° Passo: ANOVA
# 07° Passo: Estatísticas descritivas por cluster
#%%
# Carregamento da base de dados (notas_fatorial)
df = pd.read_csv('Países PCA Cluster.csv') 

#%%
# Visualização da base de dados
print(df)

#%%
# Estatísticas descritivas
summary = df.describe()
print('############ ESTATÍSTICAS DESCRITIVAS ############')
print(summary)


#%%

# INÍCIO ANÁLISE POR COMPONENTES PRINCIPAIS

#%%
# Scatter e ajuste entre as variáveis 'renda' e 'expectativa de vida'

x = df['income'].values.reshape(-1, 1)
y = df['life_expec'].values.reshape(-1, 1)
fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(1, 1, 1)
ax.scatter(x, 
           y,
           color = 'black')


ax.grid()
ax.set_xlabel('Renda',
              fontsize = 20)
ax.set_ylabel('Expectativa de vida',
              fontsize = 20)
plt.show()
#%%
# Scatter e ajuste entre as variáveis 'exports' e 'gdpp de vida'

x = df['exports'].values.reshape(-1, 1)
y = df['gdpp'].values.reshape(-1, 1)
fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(1, 1, 1)
ax.scatter(x, 
           y,
           color = 'black')

ax.grid()
ax.set_xlabel('exports',
              fontsize = 20)
ax.set_ylabel('gdpp',
              fontsize = 20)
plt.show()
#%%

# 01º Passo: Matriz de correlações de Pearson

# Coeficientes de correlação de Pearson para cada par de variáveis
def pearson(df1, df2):

    coeffmat = np.zeros((df1.shape[1], df2.shape[1]))
    pvalmat = np.zeros((df1.shape[1], df2.shape[1]))
    
    for i in range(df1.shape[1]):    
        for j in range(df2.shape[1]):        
            corrtest = pearsonr(df1[df1.columns[i]], df2[df2.columns[j]])  
    
            coeffmat[i,j] = corrtest[0]
            pvalmat[i,j] = corrtest[1]

    corr_coef = pd.DataFrame(coeffmat, columns=df2.columns, index=df1.columns)
    corr_sig = pd.DataFrame(pvalmat, columns=df2.columns, index=df1.columns)
    return corr_coef, corr_sig



corr_coef, corr_sig = pearson(df.iloc[:,1:], df.iloc[:,1:])
print('##### MATRIZ DE CORRELAÇÃO ######') #OK
print(corr_coef)
print('##### P-VALUE ######') #OK
print(corr_sig)

#%%
# Elaboração de um mapa de calor das correlações de Pearson entre as variáveis
ax = sns.heatmap(
    corr_coef, 
    vmin=-1, vmax=1, center=0,
    cmap=sns.diverging_palette(20, 220, n=200),
    square=True,
    annot=True, 
    fmt='.2f')

ax.set_xticklabels(
    ax.get_xticklabels(),
    rotation=45,
    horizontalalignment='right'
)

#%%

# 02° Passo: Teste de esferecidade de Barllet

def bartlett_sphericity(dataset, corr_method="pearson"):
    
    r"""
    
    Parameters
    ----------
    dataset : dataframe, mandatory (numerical or ordinal variables)
        
    corr_method : {'pearson', 'spearman'}, optional
        
    Returns
    -------
    out : namedtuple
        The function outputs the test value (chi2), the degrees of freedom (ddl)
        and the p-value.
        It also delivers the n_p_ratio if the number of instances (n) divided 
        by the numbers of variables (p) is more than 5. A warning might be issued.
        
        Ex:
        chi2:  410.27280642443156
        ddl:  45.0
        p-value:  8.73359410503e-61
        n_p_ratio:    20.00
        
        Out: Bartlett_Sphericity_Test_Results(chi2=410.27280642443156, ddl=45.0, pvalue=8.7335941050291506e-61)
    
    References
    ----------
    
    [1] Bartlett,  M.  S.,  (1951),  The  Effect  of  Standardization  on  a  chi  square  Approximation  in  Factor
    Analysis, Biometrika, 38, 337-344.
    [2] R. Sarmento and V. Costa, (2017)
    "Comparative Approaches to Using R and Python for Statistical Data Analysis", IGI-Global.
    
    Examples
    --------
    illustration how to use the function.
    
    >>> bartlett_sphericity(survey_data, corr_method="spearman")
    chi2:  410.27280642443145
    ddl:  45.0
    p-value:  8.73359410503e-61
    n_p_ratio:    20.00
    C:\Users\Rui Sarmento\Anaconda3\lib\site-packages\spyderlib\widgets\externalshell\start_ipython_kernel.py:75: 
    UserWarning: NOTE: we advise  to  use  this  test  only  if  the number of instances (n) divided by the number of variables (p) is lower than 5. Please try the KMO test, for example.
    backend_o = CONF.get('ipython_console', 'pylab/backend', 0)
    Out[12]: Bartlett_Sphericity_Test_Results(chi2=410.27280642443156, ddl=45.0, pvalue=8.7335941050291506e-61)
    """
    
    import numpy as np
    import math as math
    import scipy.stats as stats
#    import warnings as warnings
    import collections

    #Dimensions of the Dataset
    n = dataset.shape[0]
    p = dataset.shape[1]
#    n_p_ratio = n / p
    
    #Several Calculations
    chi2 = - (n - 1 - (2 * p + 5) / 6) * math.log(np.linalg.det(dataset.corr(method=corr_method)))
    #Freedom Degree
    ddl = p * (p - 1) / 2
    #p-value
    pvalue = 1 - stats.chi2.cdf(chi2 , ddl)
    
    Result = collections.namedtuple("Bartlett_Sphericity_Test_Results", ["chi2", "ddl", "pvalue"],  rename=False)   
    
    #Output of the results - named tuple
    result = Result(chi2=chi2,ddl=ddl,pvalue=pvalue) 

    
    #Output of the function
    # if n_p_ratio > 5 :
    #     print("n_p_ratio: {0:8.2f}".format(n_p_ratio))
    #     warnings.warn("NOTE: we advise  to  use  this  test  only  if  the number of instances (n) divided by the number of variables (p) is lower than 5. Please try the KMO test, for example.")
        
    
    return result

bartlett_chisq, bartlett_degrees, bartlett_pvalue = bartlett_sphericity(df.iloc[:, 1:], 'pearson')

print('##### CHI-SQUARED ######') #OK
print(bartlett_chisq)
print('##### DEGREES OF FREEDOM ######') #OK
print(bartlett_degrees)
print('##### P-VALUE ######') #OK
print(bartlett_pvalue)


#%%

# 03° Passo: Autovalores e autovetores da matriz de correlação de Pearson. Da pra tirar a variância compartilhada, que é a porcentagem de cada autovalor da soma deles

# Autovalores e autovetores
eigenvalues,eigenvectors = sla.eig(corr_coef)
eigenvalues = np.real(eigenvalues)

idx = eigenvalues.argsort()[::-1]   
eigenvalues = eigenvalues[idx]
eigenvectors = eigenvectors[:,idx]
eigenvectors[:, 1] = -eigenvectors[:, 1]
eigenvectors = -eigenvectors
#%%
print('##### AUTOVALORES ######') #OK
print(eigenvalues)
print('\n')
print('##### VARIÂNCIAS COMPARTILHADAS ######') #OK
eigenvalues_variance = eigenvalues/np.sum(eigenvalues)
print(eigenvalues_variance)
#%%
print('##### AUTOVETORES ######') #OK, sinais diferentes mas não deveria mudar os resultados
print(eigenvectors)

#%%

# 04° Passo: Scores fatoriais (autovetores dividido pela raiz do seu autovalor)

# Scores fatoriais 
fatoriais = eigenvectors/np.sqrt(eigenvalues[None,:]) #OK, sinais diferentes mas não deveria mudar os resultados
print(fatoriais)

#%%

# 05° Passo: Zscores (são as observações padronizadas)

# Z-scores
df_zscore = zscore(df.iloc[:,1:], ddof = 1)
print('##### Z-SCORES ######') #OK
print(df_zscore)

#%%

# 06° Passo: Fatores (multiplicação matricial dos Zscores com os scores fatoriais)

# Fatores
fatores = np.matmul(df_zscore, fatoriais)
print('##################### FATORES ####################') #OK, sinais diferentes mas não deveria mudar os resultados
print(fatores)
print('\n')
print('################# PEARSON FATORES ################') #OK
print(pearson(fatores, fatores)[0])

#%%

# 07° Passo: Cargas fatoriais (correlaçaõ de Pearson entre os fatores e as variáveis originais)

cargas_fatoriais, _ = pearson( df.iloc[:,1:], fatores)
print('################ CARGAS FATORIAIS ################') #OK, sinais diferentes mas não deveria mudar os resultados
print(cargas_fatoriais)
print('\n')
print('################# COMUNALIDADES ##################') #OK
print(np.sum(cargas_fatoriais**2, axis = 1))
print('\n')
print('################## AUTOVALORES ###################') #OK
print(np.sum(cargas_fatoriais**2, axis = 0))
print(eigenvalues)

#%%

# 08° Passo: Critério de Kaiser (autovalores maiores que 1), para definir o número de fatores
# 09° Passo: Comunalidades dos autovalores de Kaiser

# Critério de Kaiser
kaiser = np.where(eigenvalues >1)[0]
cargas_fatoriais_kaiser = cargas_fatoriais.iloc[:, kaiser.tolist()]
print('############# CARGAS FATORIAIS KAISER ############')
print(cargas_fatoriais_kaiser)
print('\n')
print('############## COMUNALIDADES KAISER ##############')
print(np.sum(cargas_fatoriais_kaiser**2, axis = 1))

#%%

# 10° Passo: Plotar as cargas fatoriais (loading plot)

# Loading plot
fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(1, 1, 1)
x =  cargas_fatoriais_kaiser.iloc[:,0]
y = cargas_fatoriais_kaiser.iloc[:,1]
ax.scatter(x, 
           y,
           color = 'black',
           s = 60)
ax.axhline(0, color = 'grey', alpha = 0.5)
ax.axvline(0, color = 'grey', alpha = 0.5)


ax.set_xlabel('Principal Component 1',
              fontsize = 20)
ax.set_ylabel('Principal Component 2',
              fontsize = 20)

ax.set_xlim(-1.1, 1.1)
ax.set_ylim(-1.1, 1.1)

text =  cargas_fatoriais_kaiser.index.values
for i in range(len(x)):
    plt.annotate(text[i], (x[i], y[i] + 0.02))
  

plt.show()
#%%
# Adicionando os fatores extraídos no banco de dados original
for i in kaiser:
    column_name = 'fator_' + str(i + 1)
    df[column_name] = fatores.iloc[:,i].values


# FIM DA ANÁLISE POR COMPONENTES PRINCIPAIS    
#%%

# INÍCIO DA ANÁLISE DE CLUSTER


#%%

# 01° Passo: Padronização das variáveis com o ZScore
 
# Análise dos fatores (média e desvio padrão)
summary = df.iloc[:, -3:].describe()
print(summary)

## ATENÇÃO: os clusters serão formados a partir dos 3 fatores
## Não aplicaremos o Z-Score, pois os fatores já são padronizados

#%%

# 02° Passo: Matriz de dissimilaridades    

# Matriz de dissimilaridades
matrix_D = squareform(pdist(df.iloc[:, -3:],
                 metric = 'euclidean'))

#%%
# Elaboração da clusterização hierárquica

# 03° Passo: Clusterização

linkage_data = sch.linkage(df.iloc[:, -3:], 
                        method = 'complete',
                        metric = 'euclidean')

#%%
# Construção do dendrograma

# 04° Passo: Dendograma

fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(1, 1, 1)
sch.dendrogram(linkage_data, 
               ax = ax,
               color_threshold=0, 
               above_threshold_color= 'black')
# sch.fcluster(linkage_data, 
#                 0, 
#                 criterion = 'monocrit', 
#                 monocrit = 5)
plt.title("Dendrograma de Cluster",
          fontsize = 20)
plt.xticks(fontsize = 20)
plt.yticks(fontsize = 20)
plt.show()

#%%

# 05° Passo: Definir número de clusters e adicioná-los no dataframe

# # Criando variável categórica para indicação do cluster no banco de dados
## O argumento 'n_clusters' indica a quantidade de clusters, faremos com 10 clusters

cutree = sch.cut_tree(linkage_data, n_clusters = 10)
if 'cluster_H' not in df:
    df.insert(len(df.columns), 
              "cluster_H" ,
              pd.DataFrame(cutree, dtype='category'))
print(df)

#%%

# 06° Passo: ANOVA

# Análise de variância de um fator (ANOVA). Interpretação do output:

## Mean Sq do cluster_H: indica a variabilidade entre grupos
## Mean Sq dos Residuals: indica a variabilidade dentro dos grupos
## F value: estatística de teste (Sum Sq do cluster_H / Sum Sq dos Residuals)
## Pr(>F): p-valor da estatística 
## p-valor < 0.05: pelo menos um cluster apresenta média estatisticamente diferente dos demais
def anova_table(aov):
    aov['mean_sq'] = aov[:]['sum_sq']/aov[:]['df']

    #aov['eta_sq'] = aov[:-1]['sum_sq']/sum(aov['sum_sq'])

    #aov['omega_sq'] = (aov[:-1]['sum_sq']-(aov[:-1]['df']*aov['mean_sq'][-1]))/(sum(aov['sum_sq'])+aov['mean_sq'][-1])

    #cols = ['sum_sq', 'df', 'mean_sq', 'F', 'PR(>F)', 'eta_sq', 'omega_sq']
    cols = ['sum_sq', 'df', 'mean_sq', 'F', 'PR(>F)']
    aov = aov[cols]
    return aov

#%%
# ANOVA da variável 'fator_1'
model = ols('fator_1 ~ C(cluster_H)', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=2)
print(anova_table(aov_table))
# ANOVA da variável 'fator_2'
model = ols('fator_2 ~ C(cluster_H)', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=2)
print(anova_table(aov_table))
# ANOVA da variável 'fator_3'
model = ols('fator_3 ~ C(cluster_H)', data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=2)
print(anova_table(aov_table))


#%%

# 07° Passo: Estatísticas descritivas por cluster

agrupar_cluster = df.groupby('cluster_H')
for i in range(len(agrupar_cluster)):
    cluster = agrupar_cluster.get_group(i)
    summary = cluster.describe()
    print('############ CLUSTER ' + str(i) +' ############')
    print(summary)

# FIM!
#%%

# APÊNDICE: PCA PELO PRINCE

#%%
# Carregamento da base de dados (notas_fatorial)
df = pd.read_csv('Países PCA Cluster.csv') 

#%%

pca = prince.PCA(n_components=3, n_iter=3, rescale_with_mean=True,rescale_with_std=True,
                 copy=True,check_input=True, engine='auto', random_state=42)

pca = pca.fit(df.iloc[:, 1:])


print(pca.row_coordinates(df.iloc[:, 1:])) # Aqui são os que vão no dataframe

print(pca.column_correlations(df.iloc[:, 1:])) # Aqui são os que vai plotar no loading plot
